### PR TITLE
support request scope

### PR DESCRIPTION
**Pull request by tarazou9**
Use await resolve() instead of get() to support request scope.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -44,14 +44,14 @@ export class DataLoaderInterceptor implements NestInterceptor {
     const ctx: any = graphqlExecutionContext.getContext();
 
     if (ctx[NEST_LOADER_CONTEXT_KEY] === undefined) {
-      ctx[NEST_LOADER_CONTEXT_KEY] = (type: string): NestDataLoader<any, any> => {
```
Comment (krislefeber) [inline index.ts]: Can the return type of this method remain specified?

File: index.ts [hunk 2]
```diff
@@ -44,14 +44,14 @@ export class DataLoaderInterceptor implements NestInterceptor {
...
+      ctx[NEST_LOADER_CONTEXT_KEY] = async (type: string) => {
         if (ctx[type] === undefined) {
           try {
-            ctx[type] = this.moduleRef
-              .get<NestDataLoader<any, any>>(type, { strict: false })
+            ctx[type] = (await this.moduleRef
+              .resolve<NestDataLoader<any, any>>(type, undefined, { strict: false }))
```
Comment (krislefeber) [inline index.ts]: I'm not very familiar with the resolve method, but by passing in a value of undefined, won't it always generate a new instance of the requested data loader class? I think we would want them to be the same per request: https://docs.nestjs.com/fundamentals/module-ref#retrieving-instances

File: index.ts (hunk 1 above)
Comment (tarazou9) [reply index.ts]: Yes! I put back the return type here. Thank you for the comment!

File: index.ts (hunk 2 above)
Comment (tarazou9) [reply index.ts]: Thank you for pointing this out! I'm able to get the request out of the graphqlExecutionContext and get a contextId. Please let me know what would think!

**Comment by krislefeber** [comment]
@tarazou9, new version published as 2.0.1. Thank you for the PR!

**Comment by krislefeber** [comment]
@tarazou9, after using this locally, I have ran into issues where the scope no longer appears to be working. Do you have specific scenarios where the scoping was not working? Until then, I am rolling back the changes.

**Comment by tarazou9** [comment]
@krislefeber  Would you please let me know what are your test scenarios and what's the error message that you are getting? I tested locally by using Requested scope and it worked for me in that case.  Would you have a test repo so that I can look into the error? And I'm sorry for introducing the error!

**Comment by krislefeber** [comment]
Within the dataloader implementation, add a logging statement for each time it runs and print out the IDs. It will log duplicate IDs with your change, while before there are no duplicates. I will try to create a small repro

**Comment by tarazou9** [comment]
Thank you! Let me try to reproduce that locally.

**Comment by tarazou9** [comment]
I ran it locally with my loader, was only able to see the context Id printed once. I will try with your repo once you have it. :)

**Comment by krislefeber** [comment]
I think the reason you're not seeing it is because you are only doing a single request to data loader in your request. The issue is when you are loading a one to one relation on a list. If you have a relation like account => address, and you have a query that looks like this, which returns multiple accounts.
`gql
query{
  getAccounts {
     id
     address {
        street_name
     }
  }
}
`
When loading the relation from account to address, it will get a new instance for each address, and not load them up all at once.

**Comment by krislefeber** [comment]
Can you share an instance of where the context is not being respected? I've triggered the same query multiple times, and it always creates a new version of the data loader per request because it is directly referenced on the dataloader instance `return ctx[NEST_LOADER_CONTEXT_KEY](data);`

**Comment by tarazou9** [comment]
@krislefeber  Would you please clarify, since I'm not quite sure that I understand your last reply? The issue that I'm trying to fix is making sure the loader is working with @Injectable({ scope: Scope.REQUEST }) So I would expect it to create a new loader per request. Would you share me with your test repo by any chance?

**Comment by krislefeber** [comment]
The loader is always going to be based on the request. You don't need to
specify the Injectable decorator, since the loader is instantiated and
tacked onto the request object.

**Comment by tarazou9** [comment]
@krislefeber  It's not the loader that's the problem, it is the request scoped service that should be injected into the data-loader is resolving to undefined. I'm trying to get a test repo out, so you can see the issue that I'm running into/resolving.

**Comment by tarazou9** [comment]
@krislefeber  Here is the test repo, to reproduce the error. https://github.com/tarazou9/test-dataloader
Below is the query to run
query {
  person(id: "12345"){
   Id,
    PersonHierarchies {
      Id
    }
  }
}

**Comment by krislefeber** [comment]
I think the problem you're running into is due to the fact that you are trying to perform a 1 to N query. Dataloader itself isn't designed for that. For every key you pass in, it must resolve to exactly one other value with a matching key: https://github.com/graphql/dataloader/blob/master/README.md#batch-function

**Comment by tarazou9** [comment]
@krislefeber the issue that caused it is because the service injected into the loader is never getting resolved, it’s a DI issue with the different scopes.

**Comment by tarazou9** [comment]
@krislefeber  I see what you meant about the loader was created multiple times now, I sent out another Pull Request to cache the context id, so the loader won't be created everytime. Also updated my "test-dataloader" repo to return more than one items, to make sure loader only created once.

**Pull request merged**